server/seeder/modules/fileComposer/fileComposer.js

```diff
@@ -0,0 +1,375 @@
+import {
+  formatString,
+  hashString,
+  getMeanFromArray,
+  getDeviationFromArray
+} from "../../helpers/utils.js";
+// import util from "util";
+
+/**
+ * Class to process XLS sheets and turn it to JS objets
+ */
+export default class FileComposer {
+  /**
+   * FileComposer constructor
+   * @param {Sheet content from fileReader} sheet
+   */
+  constructor(sheet) {
+    this.sheet = sheet;
+    this.maxDepth = 10;
+    this.floatPrecision = 3;
+  }
+
+  /**
+   * Group internal methods to process units sheet
+   */
+  processUnits() {
+    for (let i = 0, l = this.sheet.length; i < l; i++) {
+      const row = this.sheet[i];
+      row.type = row.Type;
+      row.description = row.Description;
+      delete row[`Type`];
+      delete row[`Description`];
+    }
+  }
+
+  /**
+   * Group internal methods and execute in order to process categories sheet
+   */
+  processCategories() {
+    // 1 – Update object keys from sheet headers
+    this.#updateKeys();
+    // 2 – Construct path (based on unique name)
+    this.#addPath();
+    // 3 – Add category ID (hash from fullpath)
+    this.#addCategoryId();
+    // 4 – Add reference to children categories
+    this.#addChildrens();
+    // 5 – Calculate average for parent categories
+    this.#addAverage();
+    // 6 – Clean unused keys and order its
+    this.#orderObject();
+  }
+
+  /**
+   * Construct and return a deep tree with children categories
+   */
+  #getDeepTree() {
+    const deepTree = [],
+      levels = [deepTree];
+
+    this.sheet.forEach((row) => {
+      levels[row.depth - 1].push({
+        ...row,
+        children: (levels[row.depth] = [])
+      });
+    });
+    return deepTree[0];
+  }
+
+  /**
+   * Find a node in tree by passing its uniqueName
+   * Return the node object (result) and its path (path)
+   */
+  #findInTree(name, tree) {
+    if (tree.uniqueName === name) {
+      const path = [tree.name];
+      return { result: tree, path };
+    } else if (tree.children) {
+      for (const child of tree.children) {
+        const tmp = this.#findInTree(name, child);
+        if (!(Object.keys(tmp).length === 0 && tmp.constructor === Object)) {
+          tmp.path.unshift(tree.name);
+          return tmp;
+        }
+      }
+      return {};
+    }
+  }
+
+  #updateKeys() {
+    // Loop over rows
+    for (let i = 0, l = this.sheet.length; i < l; i++) {
+      const row = this.sheet[i];
+      // Loop over level's names
+      for (let n = 0; n < this.maxDepth; n++) {
+        if (row[`Level ${n + 1}`] !== undefined) {
+          // Add title key with value from level
+          row.title = row[`Level ${n + 1}`];
+          // Add depth key from level number
+          const currentLevel = Object.keys(row)[0].replace("Level ", "");
+          row.depth = parseInt(currentLevel);
+          // Delete unused level key
+          delete row[`Level ${n + 1}`];
```

Comment:
Not sure but I would say the level is n+1
```suggestion
        const currentLevel = n+1;
        const key = `Level ${currentLevel }`;
        if (row[key] !== undefined) {
          // Add title key with value from level
          row.title = row[key];
          // Add depth key from level number
          row.depth = parseInt(currentLevel);
          // Delete unused level key
          delete row[key];
          break;
```

Reply:
row[key] https://usbeketrica.com/media/83950/download/rocky2.jpg?v=1&inline=1